### ISSUE TITLE
Debug Issue with Delete Scan Kernel on Nvidia GPU (Surface Studio)

### ISSUE THREAD

**Issue by controllerface** [controllerface/bvge#46]
When doing some tests on my laptop, I noticed that the delete scan kernel failed due to a resource issue. When I halved the max number of work items per group count, it worked fine, so this points to a possible issue with buffer allocation size.

To check this, I will probably need to verify that the buffer size that will be allocated for the scan calls is not too big, and if it is, I will need to lower the max group size to accommodate. I think the `CL_DEVICE_MAX_MEM_ALLOC_SIZE` is probably the value I need to check during init, but I am not 100% sure. May need to do some experimentation,  but hopefully this is something that can be statically determined. I don't want to have to add a flag or something to get around this on different hardware.

**Comment by controllerface** [comment]
Ok so, CL_DEVICE_MAX_MEM_ALLOC_SIZE  was not quite the right value to check, that seems to apply to global buffers, CL_DEVICE_LOCAL_MEM_SIZE  looks like it is more likely what I need. I did some real quick checks and in the delete kernel, on my laptop, the amount of memory it tries to reserve is _exactly_ this limit. Though, strictly speaking, since it is right at the limit, it probably _should_ work, it seems like it is probably safer to ensure that the max value used in a scan kernel is actually _below_ this. In any case, it should be possible to make this work programmatically.

However, I have actually found a different issue as well, related to the uniform grid size. I recently changed this to be a static size instead of being set dynamically based on the window size. It looks like I may need to revert this, as it no longer works correctly on my laptop. It seems the current setting only works if the window size is 1920 x 1080. At the moment, I don't have a monitor capable of a greater resolution than that, so I didn't notice this problem.

When I first was working on this code, I had a renderer that showed me the extents of the uniform grid, but I removed it as it was not easy to port to the newer method of doing things as I started moving all the physics code onto the GPU. It seems I will need to take another crack at that, at the very least I need to draw the extents of the grid so i can tell what part of the screen it covers. If need be, I can revert the static grid size as it was not truly _necessary_ it just made things a tad more efficient on paper. If that small improvement comes at the cost of being able to support multiple resolutions though, it isn't going to be worth it.

Perhaps I can figure out a way to compromise slightly, and just take the maximum size for the system where the program is running, and use that so i can still generate the sizes only once.

**Comment by controllerface** [comment]
OK, was able to solve the gid issue, turns out I was not actually updating the origin value correctly, I just needed to take into account the difference between the screen size and the gird size when updating the original each frame. I also fixed the local memory issue by adding some logic that checks the maximum block sizes at startup, I added a loop that halves the effective maximum until the calculated total fits into the GPU local size. This does mean that GPUs that have toot small of a local memory size will have to perform fewer scan kernel work groups than the hardware maximum, but this has no noticeable impact on my test laptop, and doesn't affect my desktop which has a local buffer with far more memory space, and does not enter the halving loop.
